### ISSUE TITLE
Package with new resolved file fails to build on older versions of Swift

### ISSUE THREAD

**Issue by stephencelis** [SwiftPackageIndex/SwiftPackageIndex-Server#3450]
**Please describe the bug**

We have a package that supports 5.9 but is not showing support on SPI due to a build failure related to Package.resolved versioning.

**Explain the steps needed to reproduce the bug**

Here's an example build failure:

> Package.resolved file is corrupted or malformed; fix or delete the file to continue: unknown 'PinsStorage' version '3' at '/Users/admin/builder/spi-builder-workspace/Package.resolved'.
GENERIC FAILURE 5.9 iOS

— https://swiftpackageindex.com/builds/8D491727-C694-4E8F-84E9-8A4CEAB091ED

**What was the expected behaviour?**

I'd expect builds to succeed, either by ignoring or deleting the resolved file.

**Comment by finestructure** [comment]
Thanks for the report, @stephencelis ! The build you linked to has been garbage collected before I had a chance to look at it. What package is this happening with?

**Comment by finestructure** [comment]
Out of curiosity: wouldn't this also affect builds in your CI? How are you dealing with it there?

**Comment by danielepantaleone** [comment]
I'm experiencing the very same issue on my swift package: https://swiftpackageindex.com/builds/7080A3DC-D844-487C-8253-C53B8632B1F3

**Comment by finestructure** [comment]
Thanks for the report, @danielepantaleone . For future reference, the package is https://swiftpackageindex.com/danielepantaleone/BlueConnect

**Comment by danielepantaleone** [comment]
Yes, I know. I just linked the build log showing the error on Swift 5.9 🙂

**Comment by finestructure** [comment]
(We delete builds when new package versions are discovered - I just wanted to make sure we record the package because when that happens your link will be a 404 :) )

**Comment by danielepantaleone** [comment]
Thanks for clarifying, I didn't know about that (still getting used to swift package index 😁)

**Comment by finestructure** [comment]
I can reproduce the error but in trying to set up a test case I'm come across something puzzling. When I run `swift package resolve` on a project with Swift 6, it creates a version 2 `Package.resolved` file:

```json
{
  "pins" : [
    {
      "identity" : "semanticversion",
      "kind" : "remoteSourceControl",
      "location" : "https://github.com/SwiftPackageIndex/SemanticVersion",
      "state" : {
        "revision" : "ea8eea9d89842a29af1b8e6c7677f1c86e72fa42",
        "version" : "0.4.0"
      }
    }
  ],
  "version" : 2
}
```

The `Package.swift` file has a tools-version 5.9 like yours. How did you managed to create a `Package.resolved` file with `"version" : 3`?

**Comment by danielepantaleone** [comment]
I believe it's due to the swift language version that I added in my swift package to enable Swift 6 language mode:

```
swiftLanguageVersions: [
    .v5,
    .version("6")
]
```

**Comment by finestructure** [comment]
That doesn't change it when I add this. Moreover, if I check out version 1.1.0 in your package and do `rm Package.resolved && swift package resolve`, I get a version 2 `Package.resolved`.

**Comment by finestructure** [comment]
Also, you'll notice in revision [a58568d0bcb79929cad2e11a30f75b20a2ad1ee1](https://github.com/danielepantaleone/BlueConnect/commit/a58568d0bcb79929cad2e11a30f75b20a2ad1ee1) where version 3 first appears in your project, the `Package.swift` file doesn't have a language mode setting.

Also there, if I recreate `Package.resolved` I get a version 2 `Package.resolved`.

I'm simply unable to create any version 3 of that file 😅

**Comment by danielepantaleone** [comment]
I may have found the cause. While adding support to Swift 6 in my package I temporarily changed `swift-tools-version` to `6.0`. Apparently, changing it back to `5.9` and enabling compatibility with Swift 6 using `swiftLanguageVersions` doesn't revert the `Package.resolved` back version 2 (unless deleted).

I pushed a branch on my swift package repository with this configuration:

https://github.com/danielepantaleone/BlueConnect/tree/swift-package-version-3

If you checkout this branch and run `rm Package.resolved && swift package resolve` you should be able to generate a version 3 of `Package.resolved`.

**Comment by finestructure** [comment]
Ah, that's it - thanks!

So in principle, we can detect and fix this. However, I believe this _is_ actually a configuration problem that will break builds in general, not just on SPI.

I'm curious about your thoughts @danielepantaleone and @stephencelis if you think this is something we should try and patch or if you consider this is an issue better fixed on the repository side.

**Comment by danielepantaleone** [comment]
I think you are right. I also believe this should be fixed in the swift package configuration. Nevertheless, it was kinda "hard/boring" tracing the issue, so maybe something more prominent can be printed in the SPI main page (even though it's not an SPI issue but it could help)?

I'm talking about something like the badge for swift 6 data safety errors:

<img width="302" alt="Screenshot 2024-10-28 at 12 35 48" src="https://github.com/user-attachments/assets/fad428d5-b029-4411-a58d-fd67f73ed7df">

**Comment by finestructure** [comment]
Unfortunately, that kind of information would be hard to feed back to where I can be displayed. The build logs would remain the place where these kind of errors surface automatically.

Honestly, if the choice were _only_ between lots of logic to parse out and display this error and just fixing it, I'd just fix it. That's way less complicated :)

**Comment by finestructure** [comment]
Ok, after speaking to Stephen I now know how this can happen in the general case. An example repository is https://github.com/pointfreeco/swift-case-paths.

It has a `Package@swift-6.0.swift` file, which is tools version 6, obviously, but also a `Package.swift` file with tools version 5.9.

This will then generate a v3 resolved file when running with Swift 6 and it makes running package resolution fail when running with Swift 5.9 due to the existing `Package.resolved` file. (FWIW, Swift 5.10 doesn't have this problem.)

This is therefore not a package problem and I can now easily recreate this scenario for a regression test. We'll work around this in our build system.

**Comment by finestructure** [comment]
Ok, we've deployed a fix for this and I've re-triggered the 5.9 builds for

- swift-case-paths
- swift-navigation
- BlueConnect

I see some builds now passing and the reported error is gone. However, the builds aren't green for the same platforms as 5.10. That might be expected but I suspect it's not. Please have a look if there are further problems and open a new issue if so!

I'll close this one in the meantime :)
